sensor/InternalScanSensor.js

```diff
@@ -40,22 +40,48 @@ const am2 = new AM2();
 const sem = require('../sensor/SensorEventManager.js').getInstance();
 const featureName = "internal_scan";
 
+const extensionManager = require('./ExtensionManager.js')
+
 class InternalScanSensor extends Sensor {
   constructor() {
     super();
   }
 
-  async run() {
+  async apiRun() {
+
     this.running = false;
     this.supportPorts = ["tcp_23", "tcp_80", "tcp_21", "tcp_3306", "tcp_6379"]; // default support: telnet http ftp mysql redis
     if (platform.getName() === 'gold') {
       this.supportPorts.push("tcp_22"); // gold: ssh
     }
-    //this.supportPorts = ["tcp_23"];
-    sem.once("DeviceServiceScanComplete", async (event) => {
-      await this.checkAndRunOnce();
+
+    extensionManager.onGet("getScanStatus", async (msg, data) => {
+      return this.scanStatus;
     });
 
+    extensionManager.onCmd("killScanSession", async (msg, data) => {
+      this.killCmd = true;
+      if (this.currentPid) {
+        const cPid = await execAsync(`ps -ef| grep nmap| awk '$3 == '${this.currentPid}' { print $2 }'`).then(result => result.stdout.trim()).catch(() => null);
+        const ccPid = await execAsync(`ps -ef| grep nmap| awk '$3 == '${cPid}' { print $2 }'`).then(result => result.stdout.trim()).catch(() => null);
+        const cccPid = await execAsync(`ps -ef| grep nmap| awk '$3 == '${ccPid}' { print $2 }'`).then(result => result.stdout.trim()).catch(() => null);
+        if (cccPid) await execAsync(`sudo kill -9 ${cccPid}`).catch((err) => { });
+      }
+    });
+
+    extensionManager.onCmd("startScanSession", async (msg, data) => {
+      if (!this.running) {
+        this.killCmd = false;
+        await this.runOnce();
+        return {"msg":"start scan"} 
+      } else {
+        return {"msg":"previous scan is running"} 
+      }
+    });
+
+  }
+
+  async run() {
```

Comment:
do not let it run automatically. This is only triggered by on-demand requests from the app.